lib/generate-critical-css.js

```diff
@@ -122,6 +124,19 @@ async function generateCriticalCSS( {
 				.filter( ( s ) => ! selectorPages[ s ].has( url ) )
 				.forEach( ( s ) => dangerousSelectors.add( s ) );
 
+			// Raise an error if there was no Critical CSS in the page. It is a good sign that a problem occurred.
+			if (
+				Object.keys( selectorPages ).length === 0 &&
+				pageSelectors.length === 0
```

Comment:
I'm not sure this check is correct / efficient. But to explain why, I think I need to break down what these variables contain - an exercise which is going to highlight just how badly they are named. ;)

`selectorPages` is an object initialized on line 87. It contains one key for every CSS selector (from CSS rules in any of the CSS files), with the Set of page URLs which include them (via CSS files they reference). It's used to identify which CSS selectors may affect each file, to help identify potentially risky rules to include in Critical CSS.

`pageSelectors` is an object collated for each page URL, which contains all of the selectors which match an element somewhere on the page.

If `selectorPages` is empty, then `pageSelectors` will always end up empty for every URL.

So, which and where we should check depends on the exact conditions we're aiming for here:

- to throw an error if there are no CSS files found, we can just check for the number of them in `cssFiles` after `collateCssFiles` finishes.
- to throw an error if there are no valid CSS Selectors found, we can just check for the number of keys in `selectorPages` after its initialized on line 87.
- to throw an error if a Page URL has no elements which match any selectors in their CSS Files, we can just check `pageSelectors` here (line 129) - but no need to also check `selectorPages` here.
- to throw an error if no Critical CSS gets generated, I would simply do it at the end if the resultant CSS is empty.

My gut says the last approach is the most correct / simplest approach, as it gets right to the heart of the problem: If no Critical CSS rules are found, something smells fishy. :)

But I'm open to discussion here. I just feel this particular set of checks isn't quite right, because it's a mix of two approaches.

Reply:
hi @thingalon,

Thanks for the clarification. it is very useful.

My intention to put  the logic in the loop to be able to tight easily the error to at least one URL without having to do something like `SuccessTargetError` error with child errors so once the error is reported on the frontend, then by clicking on the url, the user will show straightaway the issue of the page without the CSS.

I will look at doing some refactoring to support your last approach but being able to capture the urls (with the critical CSS parameters) maybe in a similar fashion of the `SuccessTargetError` with child errors.